### PR TITLE
Add support for distributed functions in extensions of distributed actors

### PR DESCRIPTION
**Pull request by DougGregor**
Fix a few minor issues in the type checker and SILGen to properly cope with
distributed functions defined within extensions of distributed actors.
While here, centralize the logic that adds the "_remote_" function.

... and since this uncovered a problem with the mangling for distributed
functions, fix that, too.

Fixes rdar://84325525.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

File: lib/Sema/CodeSynthesisDistributedActor.cpp [hunk 1]
```diff
@@ -189,7 +188,24 @@ AbstractFunctionDecl *TypeChecker::addImplicitDistributedActorRemoteFunction(
...
+  auto &C = func->getASTContext();
+  DeclContext *DC = func->getDeclContext();
+
+  // not via `ensureDistributedModuleLoaded` to avoid generating a warning,
+  // we won't be emitting the offending decl after all.
+  if (!C.getLoadedModule(C.Id_Distributed))
+    return nullptr;
+
+  return addImplicitDistributedActorRemoteFunction(DC, func);
+}
```
Comment (ktoso) [inline lib/Sema/CodeSynthesisDistributedActor.cpp:211]: thanks for the cleanup here

**Review by ktoso — APPROVED** [review]
Looks good, thanks a lot!

File: docs/ABI/Mangling.rst [hunk 2]
```diff
@@ -217,6 +217,7 @@ types where the metadata itself has unknown layout.)
   global ::= global 'To'                 // swift-as-ObjC thunk
   global ::= global 'TD'                 // dynamic dispatch thunk
   global ::= global 'Td'                 // direct method reference thunk
+  global ::= global 'TE'                 // distributed actor thunk
```
Comment (ktoso) [inline docs/ABI/Mangling.rst:220]: okey E then! Thanks for spotting this

Comment (ktoso) [reply docs/ABI/Mangling.rst:220]: Hmmm we also used the `Td` in:

```

  case Kind::DistributedThunkAsyncFunctionPointer: {
    std::string Result = getSILDeclRef().mangle();
    Result.append("Td");
    Result.append("Tu");
    return Result;
  }

```

which this doesnt fix yet -- i think there we also need to use TE right? I'll follow up with this.

**Review by ktoso — APPROVED** [review]

**Comment by ktoso** [comment]
@swift-ci please build toolchain macOS

**Comment by ktoso** [comment]
Actual failure, I think I know how to solve -- on it.

**Comment by ktoso** [comment]
Replaced by https://github.com/apple/swift/pull/39787 where the only additional fix is https://github.com/apple/swift/pull/39787/commits/e1ae0be2dd1b45fd5aec75a033338279e2b89b83

**Pull request closed without merging**